File: backend/app/schemas/admin_email.py
"""Request and response models for the email delivery log (spec 055)."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.models.email import EmailKind, EmailStatus


class DeliveryResponse(BaseModel):
    id: UUID
    kind: EmailKind
    to_email: str
    user_id: UUID | None
    #: "The relay accepted it", never "it arrived" — SMTP offers no delivery
    #: confirmation, so the UI says *accepted by relay*, which is the true thing.
    status: EmailStatus
    error_type: str | None
    duration_ms: int | None
    created_at: datetime


class DeliveryPageResponse(BaseModel):
    total: int
    entries: list[DeliveryResponse]


class MailStatusResponse(BaseModel):
    configured: bool
    window_minutes: int
    sent: int
    failed: int
    failure_rate: float
    #: Worth a banner. False below a minimum volume: one failure out of one is
    #: 100% and means nothing.
    degraded: bool

File: frontend/src/api/adminEmail.ts
import { api } from "./client";

/** The email delivery log (spec 055). */

export type EmailKind = "magic_link" | "approval_notice" | "test";
/** "The relay accepted it", never "it arrived" — SMTP confirms nothing. */
export type EmailStatus = "sent" | "failed" | "not_configured";

export interface Delivery {
  id: string;
  kind: EmailKind;
  to_email: string;
  user_id: string | null;
  status: EmailStatus;
  error_type: string | null;
  duration_ms: number | null;
  created_at: string;
}

export interface DeliveryPage {
  total: number;
  entries: Delivery[];
}

export interface MailStatus {
  configured: boolean;
  window_minutes: number;
  sent: number;
  failed: number;
  failure_rate: number;
  /** Worth a banner. False below a minimum volume. */
  degraded: boolean;
}

export interface DeliveryFilters {
  kind?: EmailKind;
  status?: EmailStatus;
  user_id?: string;
  search?: string;
  limit?: number;
}

export const getMailStatus = () => api.get<MailStatus>("/admin/email/status");

export const listDeliveries = (filters: DeliveryFilters = {}) => {
  const query = new URLSearchParams();
  for (const [key, value] of Object.entries(filters)) {
    if (value !== undefined && value !== "") query.set(key, String(value));
  }
  const suffix = query.toString();
  return api.get<DeliveryPage>(`/admin/email/deliveries${suffix ? `?${suffix}` : ""}`);
};

/** Always to the caller's own address; there is no recipient to supply. */
export const sendTestEmail = () => api.post<{ message: string }>("/admin/email/test", {});
